server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParameters.java

```diff
@@ -92,22 +96,41 @@ public enum Metric {
         REPOSITORIES("repositories"),
         ALLOCATIONS("allocations");
 
-        private String metricName;
+        private static final Map<String, Metric> metricMap = Arrays.stream(values())
+            .collect(toUnmodifiableMap(Metric::metricName, Function.identity()));
+        private final String metricName;
+
+        public static final Set<Metric> ALL = Arrays.stream(values()).collect(toUnmodifiableSet());
+        public static final Set<String> ALL_NAMES = metricMap.keySet();
 
         Metric(String name) {
             this.metricName = name;
         }
 
+        public static boolean isValid(String name) {
+            return metricMap.containsKey(name);
+        }
+
+        public static Metric get(String name) {
+            return metricMap.get(name);
+        }
+
+        public static Metric readFrom(StreamInput in) throws IOException {
+            return Metric.get(in.readString());
+        }
```

Comment:
Just to confirm, are we sure we never serialize unexpected metric?

I am thinking about a scenario when old node serialized unexpected metric name, than then upgraded node will deserialize it as null.

Reply:
Good point! I missed that. I believe current behaviour is to ignore unknown metrics in transport. I will update code and add test for this.